app/eth1wrap/client.go

```diff
@@ -0,0 +1,110 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package eth1wrap
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/obolnetwork/charon/app/errors"
+	erc1271 "github.com/obolnetwork/charon/app/eth1wrap/generated"
+)
+
+//go:generate abigen --abi=build/IERC1271.abi --pkg=erc1271 --out=generated/erc1271.go
+
+var erc1271MagicValue = [4]byte{0x16, 0x26, 0xba, 0x7e}
+
+// NewEth1Client returns a initiliazed eth1 JSON-RPC client.
+func NewEth1Client(addr string) (*Client, error) {
```

Comment:
Typically, most implementations that used eth1 client, provide `Run(context.Context)` function that can be running on a goroutine. Entire lifecycle is implemented in this function:

```
func (c *Client) Run(ctx context.Context) error {
   nativeClient, err := c.client.DialContext(ctx, c.url)
   if err != nil { ... }
   defer nativeClient.Close()
   
   for {
      // wait for context cancellation, or a disconnect event, then reconnect.
   }
}
```

Reply:
In this case you don't need any fancy "initialization".
For `erc1271.IsValidSignature(nil, hash, sig)` you need check if it is thread-safe, otherwise use a mutex to serialize possible concurrent request to verify signature.

Reply:
Maybe providing "lazy" behavior is a premature optimization. I think if the user supplied RPC urls, it expects that charon will do connect. This way we delegate this responsibility to the user. Having one extra connection to RPC and one goroutine serving it is a minor impact IMO.